### PR TITLE
feat: #1145 use cached access for applications list

### PR DESCRIPTION
**Pull request by ianliuwk1019**

### CODE REVIEW

File: frontend/src/store/FamLoginUserState.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        )
+    }
+};
+
+/**
+ * @param adminRole three admin levels: 'FAM_ADMIN', 'APP_ADMIN', 'DELEGATED_ADMIN'
+ * @returns array of applications the login user has been granted for
+ *          administering on based on adimn level.
+ */
+const getApplicationsAdministeredByAdminRole = (
```
Comment (MCatherine1994) [inline frontend/src/store/FamLoginUserState.ts:56]: Hi Ian, just personal thoughts... I wonder if we can just call it getAppsByAdminRole.... the applications administered is too long I feel....

Comment (ianliuwk1019) [reply frontend/src/store/FamLoginUserState.ts:56]: Sure, actually I like the short name initially but was afraid it is not so clear to others.
I will use the shorter name. 
I created all methods that are related and close to the state file to provide functionalities and would be easier to know where they are first before even touch components but either places are fine to me. I will move to the component.

Comment (MCatherine1994) [reply frontend/src/store/FamLoginUserState.ts:56]: oh sorry I just saw you use it in getApplicationsUserAdministers method as well, it confused me a bit.... I might prefer to separate it, so this call will just be used by fam admin

Comment (ianliuwk1019) [reply frontend/src/store/FamLoginUserState.ts:56]: Yes, those methods are useful as they are very closed related to the state. That's why it is much clear to me to be in the container of the same file and just provide them for whoever needs them (even just for one component, much like util may just be used for one function).
If we separate them out into components then component will be packed with complicated code that they are not suppose to be, and basic common logic will also need to be repeated in both places.

File: frontend/src/store/FamLoginUserState.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+ *
+ * If the way our forntend works is based on admin level to display
+ * management table instead of selecting "applicatoin" first, it won't need
+ * this particular function.
+ */
+const getApplicationsUserAdministers = () => {
+    const userAdminRoleGroups = getUserAdminRoleGroups();
+    let famApp;
+    const applicationList: Array<FamApplicationDto> = [];
+    if (userAdminRoleGroups) {
```
Comment (MCatherine1994) [inline frontend/src/store/FamLoginUserState.ts:85]: so maybe we can just go through the access array instead of go through the userAdminRoleGroups and call `getApplicationsAdministeredByAdminRole`, I feel it's a little bit hard to follow

File: frontend/src/store/FamLoginUserState.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    // clean up local storage for selected application
+    localStorage.removeItem(CURRENT_SELECTED_APPLICATION_KEY);
+};
+
+/**
+ * To refresh User granted privileges when neccssary.
+ * FamLoginUser contains a property "accesses" that sometimes needs to be
+ * refreshed and stored into localStorage at FamLoginUser object.
+ */
+const refreshCachedUserAccess = async () => {
```
Comment (MCatherine1994) [inline frontend/src/store/FamLoginUserState.ts:143]: maybe we shouldn't call it refreshCached anymore, cause we only call it when we login, and there won't be any cached access

File: frontend/src/store/FamLoginUserState.ts (hunk 1 above)
Comment (MCatherine1994) [reply frontend/src/store/FamLoginUserState.ts:56]: I'm fine to keep it here, but maybe restrict it to call used by fam admin, and in the getApplicationsUserAdministers method, just go through the whole access list, and check `access.grants.forEach`, then you don't need to reuse this method, and feel it's more clear

**Comment by MCatherine1994** [comment]
Thanks Ian! Looks great! Just some small comment for simplify and clarification. Also maybe Nick @NickSaglioni if you want to check on this pr, Ian is getting the application options inside the component now instead of preparing it in the router handler, just regarding of your question this morning. Thanks so much!

File: frontend/src/store/FamLoginUserState.ts (hunk 2 above)
Comment (ianliuwk1019) [reply frontend/src/store/FamLoginUserState.ts]: Good suggestion, although I was reusing other function for convenience but the extra mapping is unnecessary. the logic refactored.

File: frontend/src/store/FamLoginUserState.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            const accessGrants = access.grants;
+            if (access.auth_key == AdminRoleAuthGroup.FamAdmin) {
+                famApp = accessGrants.filter(
+                    (grant) => grant.application.name == FAM_APPLICATION_NAME
+                )[0].application;
+            } else {
+                accessGrants.forEach((grant) => {
+                    const app = grant.application
+                    let isNewItem = true
+                    for (let item of applicationList) {
```
Comment (MCatherine1994) [inline frontend/src/store/FamLoginUserState.ts:92]: Hi Ian, just wondering maybe we check if length of applicationList.filter((addedApp) => addedApp.id == app.id) > 0?  so we have less code? it doesn't impact functionality, I'll approve it, thanks!!

**Review by MCatherine1994 — APPROVED** [review]

File: frontend/src/store/FamLoginUserState.ts (hunk 4 above)
Comment (ianliuwk1019) [reply frontend/src/store/FamLoginUserState.ts:92]: Hi Catherine, I think it is fine, I don't always use short-form, sometimes I favor for-loop for clearness.
Thanks Catherine! I will merge so others can resolve their conflict.
